### PR TITLE
libobs: Add keyframe aligned encoder groups

### PR DESCRIPTION
**Pull request by palana**
### Description
This is meant to allow grouping encoders for e.g. eRTMP multitrack video (<=> Twitch Enhanced Broadcasting) such that IDR frames are aligned for seamless rendition switching when transmuxing the overall output stream into e.g. HLS or similar multi rendition formats

### Motivation and Context
Without this change, encoders aren't guaranteed to start on the same frame even if started "at the same time" due to encoder startup not being synchronized with the video thread; e.g. in practice we've observed a one frame offset that sometimes happens between encoders, though the offset can be arbitrarily high potentially due to lack of synchronization

With this there's explicit synchronization, so all encoders start on the same frame (and get the "same stream" of frames fed to them)

### How Has This Been Tested?
This code has been deployed in the Twitch Enhanced Broadcasting beta since January, and its effects can be observed when disabling the check in `receive_video` while using multitrack video outputs (and logging which input frame is first sent to encoders)

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by WizardCM** [comment]
Please update the relevant documentation.

**Comment by palana** [comment]
> Please update the relevant documentation.

Updated `reference-outputs.rst` in https://github.com/obsproject/obs-studio/pull/10349#event-12074277391

**Review by RytoEX — CHANGES_REQUESTED** [review]
You can omit the `docs` prefix from these commits to shorten the commit message subjects. It will be apparent from the changed files that docs were changed. If you want to draw attention to it, mention in the commit message body that "this commit also updates/modifies docs as needed" or some such.

**Comment by palana** [comment]
> You can omit the `docs` prefix from these commits to shorten the commit message subjects. It will be apparent from the changed files that docs were changed. If you want to draw attention to it, mention in the commit message body that "this commit also updates/modifies docs as needed" or some such.

Removed the `docs` prefix in https://github.com/obsproject/obs-studio/pull/10349#commits-pushed-273c2fd

**Comment by tt2468** [comment]
I have one primary concern with this PR, which is that the currently supported behavior of sharing an encoder between multiple outputs, and starting those outputs at different times, becomes restricted with this addition.

Say I have encoder A, encoder B, output 1, and output 2. I apply encoder A and B to output 1, then I apply encoder A to output 2.

I can start output 1 first, then start output 2. I can also start output 2 first, and then start output 1. Say though, with this change, I "group" encoders A/B together. In this case, I **have** to start output 1 before output 2 will finish initializing encoders.

Having this be a limitation of aligning encoder keyframes, to me, is an important issue, as any future multi-destination streaming features which intend to support enhanced broadcast will need to impose this limitation on the user.

Rodney brought up that perhaps we could take control of the GOP and manually trigger keyframes. I think that this might be an adequate, and more reliable, method of synchronizing keyframes across an "encoder group." I suppose in theory, if an encoder is "grouped", it would disable any automatic keyframe generation in the implementation, and rely on signals from libobs to produce new keyframes. This could possibly be implemented as a flag in a "flags" parameter in the encode callbacks.

File: docs/sphinx/reference-outputs.rst
Comment (palana) [reply docs/sphinx/reference-outputs.rst:770]: @RytoEX: should I remove the "the other output" bit? the output doesn't necessarily matter, all grouped encoders need to be started for any of them to actually receive frames

**Comment by derrod** [comment]
> Rodney brought up that perhaps we could take control of the GOP and manually trigger keyframes. I think that this might be an adequate, and more reliable, method of synchronizing keyframes across an "encoder group." I suppose in theory, if an encoder is "grouped", it would disable any automatic keyframe generation in the implementation, and rely on signals from libobs to produce new keyframes. This could possibly be implemented as a flag in a "flags" parameter in the encode callbacks.

To clarify, what I did is simply adding a means of forcing an IDR for resynchronisation, the encoder is still managing it by itself otherwise. Based on what vendors have told us, using an open/infinite GOP and external IDR control will also have a negative impact on quality, so it's not advised. Besides that, not all encoders even support this (e.g. VideoToolbox).

> I can start output 1 first, then start output 2. I can also start output 2 first, and then start output 1. Say though, with this change, I "group" encoders A/B together. In this case, I have to start output 1 before output 2 will finish initializing encoders.

I'm not sure there's an elegant way to solve this, the resync mechanism I have build would *technically* work in this case. You just have to let the encoder B start out of sync, and then wait for the resync to happen before starting data capture on output 1, or accept out-of-sync packets for a start ("just" in this case probably means a lot of complexity to track that), However, that would result in output 2 potentially receiving a premature IDR frame.

**Comment by palana** [comment]
> I have one primary concern with this PR, which is that the currently supported behavior of sharing an encoder between multiple outputs, and starting those outputs at different times, becomes restricted with this addition.
> 
> Say I have encoder A, encoder B, output 1, and output 2. I apply encoder A and B to output 1, then I apply encoder A to output 2.
> 
> I can start output 1 first, then start output 2. I can also start output 2 first, and then start output 1. Say though, with this change, I "group" encoders A/B together. In this case, I **have** to start output 1 before output 2 will finish initializing encoders.
> 
> Having this be a limitation of aligning encoder keyframes, to me, is an important issue, as any future multi-destination streaming features which intend to support enhanced broadcast will need to impose this limitation on the user.
> 
> Rodney brought up that perhaps we could take control of the GOP and manually trigger keyframes. I think that this might be an adequate, and more reliable, method of synchronizing keyframes across an "encoder group." I suppose in theory, if an encoder is "grouped", it would disable any automatic keyframe generation in the implementation, and rely on signals from libobs to produce new keyframes. This could possibly be implemented as a flag in a "flags" parameter in the encode callbacks.

At this point, we’ve tested & validated that this set of changes meet the requirements for the Twitch Enhanced Broadcasting use case. If there are additional use cases with new/different requirements, we’ll continue to iterated and bring new PRs that build on this after it gets merged

**Comment by palana** [comment]
@RytoEX: squashed and rebased commits just now

**Comment by Lain-B** [comment]
For the shared encoder situation (i.e. hypothetically using one of the encoders on another output and then starting that output before the output with the group), there are a few ways to handle it, but right now, these encoders will not be shared for the time being (e.g. sharing recording encoder with stream encoder) when using multi-encode. They spawn separate encoders in the enhanced broadcast implementation for the time being.

Here are some proposals to handle the keyframe alignment issue when sharing sub-encoders between multiple outputs in the future if we decide to implement it (which would be ideal at some point):

1. Drop the keyframe alignment requirement and leave it to the service to solve this in another way, likely client-side. This would be my ideal solution if it's possible to do. However, unless we do this soon, we would have to of course change the name after merge (which is annoying but fine if we have to)

2. If removing alignment requirement is not reasonable/doable, the most simple alternative is to start all encoders in a group any time one of its encoders are started up individually. This is not fantastic because if you just want to record then you have to start up many encoders no matter what, but has the benefit of being a change that should in theory be relatively trivial to implement.

3. Another proposal I've been thinking about is when starting up other encoders within an encoder group later, you could probably add somewhat complex handling to discard raw frames and make newly-started encoders wait before processing their first starting frame in order to match the precise timing of keyframes of already-started encoders. I feel like this should in theory be quite possible to do. You also might be able to discord both encoded frames *and* raw frames together as a hack in order to compensate for unknown encoder frame delay. I definitely feel this is quite possible now that I think about it.

4. There is also the proposal to manually signal keyframes ourselves rather than leave it to the encoder, which you had proposed tt. This is also a complex proposal and I am not sure all encoders among all codecs (current and upcoming) support manual keyframe placement. Plus I'm not sure how manual keyframe placement impacts encoders or how they're implemented among encocers. Probably just as complicated as solution 3 however.

Anyway, while keyframe alignment is a requirement, there are no perfect solutions. I would likely opt for my proposal of just starting all encoders in a group if one of them is started elsewhere just simply for the sake of simplicity for the time being but that's just me.

I am fairly confident however that API changes -- outside of the first proposal -- would likely not be necessary regardless of how we choose to handle it (if we choose to handle it).

**Review by Lain-B — APPROVED** [review]
There are more than one ways to adapt this later on without modifying the API. Right now I feel relatively good with it and would prefer not to stall this indefinitely. I feel fairly strongly that it can be adapted without necessitating API breakage if we want to add encoder sharing later (except if the alignment requirement is removed, which I'm guessing at this point in time is unlikely).

File: libobs/obs-output.c
Comment (RytoEX) [inline libobs/obs-output.c:2014]: Nit.
```suggestion
/* Check whether keyframes are emitted from all grouped encoders, and log
 * if keyframes haven't been emitted from all grouped encoders.
 */
```

Or
```suggestion
/* Check whether keyframes are emitted from all grouped encoders, and log
 * if keyframes haven't been emitted from all grouped encoders. */
```

**Review by RytoEX — CHANGES_REQUESTED** [review]
Nit only, which need not block merging if it doesn't get changed. Otherwise, I trust @Lain-B 's review and assessment.

File: libobs/obs-output.c
Comment (palana) [reply libobs/obs-output.c]: the second suggestion seems more common in libobs code, so going with that

File: libobs/obs-output.c
Comment (RytoEX) [inline libobs/obs-output.c:2040]: Just a note for the future: LOG_WARNING may be too noisy. We'll have to evaluate as this gets used in the wild if this should be LOG_DEBUG.

**Review by RytoEX — COMMENTED** [review]
Leaving a general comment/question.

Overall this looks okay to me, and has @Lain-B 's approval. So pending an answer to this question, this should be able to be merged. If there are still pending questions/concerns after that, particularly about exported functions, we can continue to change them until we ship a stable release with those functions.

File: libobs/obs-output.c
Comment (Lain-B) [reply libobs/obs-output.c:2040]: This message doesn't get output normally, it's basically considered an error if this error message is output, so it's probably important and makes sense to output it as a general log warning in this case. Otherwise we'll never know about it from user logs.

Comment (Lain-B) [reply libobs/obs-output.c:2040]: (Or rather considered a warning that something is going awry and should be examined. Stream should still technically work I believe)

Comment (Lain-B) [reply libobs/obs-output.c:2040]: Huh, Ryan sent me a log file where the user had dynamic bitrate enabled (which I assume isn't supported) and it *does* get very verbose with multiple messages per second. Hmm.

Comment (Lain-B) [reply libobs/obs-output.c:2040]: Well I think that's being patched so I think it'll be fine for now. If we have any issues with logging we can adjust it later.

Comment (RytoEX) [reply libobs/obs-output.c:2040]: My comment was mostly, "Let's be mindful of this and keep an eye out as this code makes it into use in the wild." I wanted it noted that I raised the concern as something to keep an eye on, but not a blocker at this time.

**Review by RytoEX — APPROVED** [review]
If there's anything else that needs fixing here, it can be fixed in follow-ups, or we can adjust as needed on our own.

File: docs/sphinx/reference-outputs.rst
Comment (RytoEX) [reply docs/sphinx/reference-outputs.rst]: I do not have a strong opinion on this, and I don't want to block further. In the future, if we find that the docs lack clarity, we can clarify them then.

**Pull request merged**
